### PR TITLE
INFINITY-2975 Make plan status behave better in a race

### PR DESCRIPTION
**Pull request by nickbp**
- Fetch plan status after phase/step status. If the steps change in the middle of this, then we end up with e.g. a plan that's still `IN_PROGRESS` when its steps are all `COMPLETE`. This is less ugly for users.
- Move some Element-level functions into internal utilities. Started on this in an earlier version of this fix, may as well keep it around.

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/types/PlanInfo.java [hunk 1]
```diff
@@ -23,11 +22,16 @@
...
+        // plan. By fetching phase statuses after the plan status, inconsistencies should typically appear as e.g. a
+        // plan that's IN_PROGRESS when the steps or phases are COMPLETE. If we did the plan status last, then we'd risk
+        // getting the opposite of that, which is less intuitive to an end user.
+        Status planStatus = plan.getStatus(); // plan status first ...
+        List<PhaseInfo> phaseInfos = plan.getChildren().stream() // ... then phases
                 .map(phase -> PhaseInfo.forPhase(phase))
                 .collect(Collectors.toList());
-
-        return new PlanInfo(phaseInfos, plan.getStrategy().getName(), plan.getErrors(), plan.getStatus());
+        return new PlanInfo(phaseInfos, plan.getStrategy().getName(), plan.getErrors(), planStatus);
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/types/PlanInfo.java]: `plan.getErrors()` also visits the children. Could a similar ordering issue not occur when displaying an `ERROR` vs `COMPLETED` state?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 2]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
...
+            result = Status.IN_PROGRESS;
+            LOGGER.debug("({} status={}) At least one child has status: {}", parentName, result, Status.PREPARED);
+        } else if (anyMatch(Status.WAITING, candidateStatuses)) {
+            result = Status.WAITING;
+            LOGGER.debug("({} status={}) At least one candidate has status: {}", parentName, result, Status.WAITING);
+        } else if (anyMatch(Status.IN_PROGRESS, candidateStatuses)) {
+            result = Status.IN_PROGRESS;
+            LOGGER.debug("({} status={}) At least one candidate has status: {}",
+                    parentName, result, Status.IN_PROGRESS);
+        } else if (anyMatch(Status.COMPLETE, childStatuses) && anyMatch(Status.PENDING, candidateStatuses)) {
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: Could we not replace all of the `else if (anyMatch(Status.COMPLETE, childStatuses) && ` conditionals here with:
```java
} else if (anyMatch(Status.COMPLETE, childStatuses)) {
    result = Status.IN_PROGRESS;
}
```
(as we already know that `(allMatch(Status.COMPLETE, childStatuses))` is false at this point)

Comment (elezar) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: (I could be missing something in the relationship between `childStatuses` and `candidateStatuses` here.)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/types/PlanInfo.java (hunk 1 above)
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/http/types/PlanInfo.java]: It is very unlikely (impossible in today's world) that the errors will ever change, but it's in principle the same kind of mistake we're trying to fix here.  A plan's status and errors should probably be grabbed as close together as possible.  There's still a race there.  Grab the errors first, then the status could produce non-empty error list and _non-ERROR_ status.   Grab the status first and get an ERROR status, but then grab errors and get an _empty_ ERROR list is also possible.  I think there's probably a few other bad combinations.  We're punting on correct consistency anyway, so I'm not too troubled either way.  None of the scenarios here seem more or less intuitive to me.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 3]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
...
+     *
+     * @param parentName the name of the parent element, used for logging
+     * @param childStatuses the statuses of the parent element's children
+     * @param candidateStatuses the statuses of the parent element's candidate children, as selected by the parent
+     *      element's Strategy
+     * @param errors any errors from the parent element
+     * @param isInterrupted whether the parent element is interrupted
+     * @return the Status to display for the parent element
+     */
+    public static <C extends Element> Status getParentStatus(
```
[the remark this answers is not in the document]
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: I like `getAggregateStatus`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 4]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
...
+
+        return result;
+    }
+
+    private static boolean allMatch(Status status, Collection<Status> statuses) {
+        return statuses.stream().allMatch(s -> s == status);
+    }
+
+    private static boolean anyMatch(Status status, Collection<Status> statuses) {
+        return statuses.stream().anyMatch(s-> s == status);
```
[the remark this answers is not in the document]
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: I think this is an incorrect use of `==`

**Comment by nickbp** [comment]
I think it's better to have it in this direction for two reasons:
- It doesn't result in producing a state that's invalid. It should be illegal for a plan to be complete when its contents are not complete.
- Users (and our tests) can safely assume that if the overall plan status is complete, then the contents are complete. Today they'd need to recurse into steps to be sure.
Making the plans perfectly consistent/locking could solve this too, but it'd be a much larger and further-reaching project. This at least gets us past the above two issues without needing to overhaul plan access.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 5]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
 
         return plan.getChildren().stream()
                 .flatMap(phase -> phase.getChildren().stream())
-                .filter(step -> step.isAssetDirty() && step.getPodInstanceRequirement().isPresent())
+                .filter(step -> (step.isPrepared() || step.isStarting())
```
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: This was the only code calling `Step.isAssetDirty()`, so to reduce the size of the `Step` interface I just moved the content of that call to here.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: Renamed to `getAggregateStatus`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 6]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
...
+     * @param isInterrupted whether the parent element is interrupted
+     * @return the Status to display for the parent element
+     */
+    public static <C extends Element> Status getParentStatus(
+            String parentName,
+            Collection<Status> childStatuses,
+            Collection<Status> candidateStatuses,
+            Collection<String> errors,
+            boolean isInterrupted) {
+        // Ordering matters throughout this method.  Modify with care.
```
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: Now that it's broken out, it's much easier to test. Added some to `PlanUtilsTest`.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java (hunk 2 above)
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: I think this is specifically checking that:
- The children contain a COMPLETE status
- The subset of children _which are candidates for doing work_ contain a PENDING status

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java [hunk 7]
```diff
@@ -76,8 +77,114 @@ public static boolean assetConflicts(PodInstanceRequirement asset, Collection<Po
...
+                    parentName, result, Status.COMPLETE, Status.PENDING);
+        } else if (anyMatch(Status.COMPLETE, childStatuses) && anyMatch(Status.STARTING, candidateStatuses)) {
+            result = Status.IN_PROGRESS;
+            LOGGER.debug("({} status={}) At least one child has status '{}' and at least one candidate has status '{}'",
+                    parentName, result, Status.COMPLETE, Status.STARTING);
+        } else if (anyMatch(Status.COMPLETE, childStatuses) && anyMatch(Status.STARTED, candidateStatuses)) {
+            result = Status.IN_PROGRESS;
+            LOGGER.debug("({} status={}) At least one child has status '{}' and at least one candidate has status '{}'",
+                    parentName, result, Status.COMPLETE, Status.STARTING);
+        } else if (!candidateStatuses.isEmpty() && anyMatch(Status.PENDING, candidateStatuses)) {
```
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanUtils.java]: Yes it is, removed the isEmpty check

**Pull request merged**